Use container Mounts API for Swarm containers.

cpuguy83 opened the pull request:
Instead of converting nicely typed service mounts into untyped `Binds`
when creating containers, use the new `Mounts` API which is a 1-1
mapping between service mounts and container mounts.

cpuguy83 commented on daemon/cluster/executor/container/validate.go:
```diff
@@ -1,43 +0,0 @@
-	"github.com/docker/swarmkit/api"
-)
-
-func validateMounts(mounts []api.Mount) error {
```
The required validation will happens in the engine now when using the new API.

cpuguy83 replied:
I'm not sure I understand why the agent would do the validation when the underlying daemon is already going to do it.

thaJeztah commented:
s/contianer/container/ in your commit message? :joy:

stevvooe reviewed (changes requested):
The validation of mounts cannot be removed.

cpuguy83 committed:
Use container Mounts API for Swarm containers.

Instead of converting nicely typed service mounts into untyped `Binds`
when creating containers, use the new `Mounts` API which is a 1-1
mapping between service mounts and container mounts.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
Rebased and added the swarm executor mount validation back.
I still don't think we need this in the executor since now we have validation in two places, but certainly won't hurt things (other than `os.Stat` being relatively expensive).

AkihiroSuda commented:
can you convert tmpfs as well? https://github.com/docker/docker/pull/26837

cpuguy83 commented:
@AkihiroSuda Sure, updated.

AkihiroSuda commented:
Are we going to have this fix in 1.13?
@cpuguy83 @stevvooe

thaJeztah commented:
@AkihiroSuda @cpuguy83 @stevvooe @vieux I added this to the board https://github.com/docker/docker/projects/1 for discussion

stevvooe commented:
LGTM

Don't we need this for 1.13?

cpuguy83 commented:
I don't think there's anything strictly depending on it for 1.13, it's just "better".

cpuguy83 commented:
This is rebased.

AkihiroSuda commented on volume/volume_linux.go:
```diff
@@ -11,7 +11,6 @@ import (
 // ConvertTmpfsOptions converts *mounttypes.TmpfsOptions to the raw option string
 // for mount(2).
-// The logic is copy-pasted from daemon/cluster/executer/container.getMountMask.
 // It will be deduplicated when we migrated the cluster to the new mount scheme.
```
please remove L14 as well? :smile:

AkihiroSuda commented on integration-cli/docker_cli_service_create_test.go:
```diff
@@ -103,3 +120,53 @@ func (s *DockerSwarmSuite) TestServiceCreateWithSecretSourceTarget(c *check.C) {
+
+func (s *DockerSwarmSuite) TestServiceCreateMountTmpfs(c *check.C) {
+	d := s.AddDaemon(c, true, true)
+	out, err := d.Cmd("service", "create", "--mount", "type=tmpfs,target=/foo", "busybox", "sh", "-c", "mount | grep foo && sleep 60")
```
nit: `tail -f /dev/null` would be better than `sleep 60`
Also, I think we should remove this service on exit

AkihiroSuda commented on integration-cli/docker_cli_service_create_test.go:
```diff
@@ -8,14 +8,15 @@ import (
 func (s *DockerSwarmSuite) TestServiceCreateMountVolume(c *check.C) {
 	d := s.AddDaemon(c, true, true)
-	out, err := d.Cmd("service", "create", "--mount", "type=volume,source=foo,target=/foo", "busybox", "top")
+	out, err := d.Cmd("service", "create", "--mount", "type=volume,source=foo,target=/foo,volume-nocopy", "busybox", "top")
```
I think we should remove this service on exit

cpuguy83 replied:
Can you explain what you mean?
Everything should be cleaned up when the test is finished.

AkihiroSuda replied:
Uh, sorry, please ignore. I was just misthinking 😓

tonistiigi commented:
LGTM

thaJeztah commented:
All :green:
